keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/TaskTrackingRepository.kt

```diff
@@ -0,0 +1,14 @@
+package com.netflix.spinnaker.keel.persistence
+
+interface TaskTrackingRepository {
+
+  fun store(task: TaskRecord)
+  fun getTasks(): Set<TaskRecord>
```

Comment:
Probably add a `limit` parameter to this like we do for most other `get` repository methods?

Reply:
Actually, come to think of it, if you go ahead with a separate monitoring loop to look at orca tasks as you have proposed, I think you would need some form of distributed coordination across the keel instances so that they’re not looking at the same tasks and/or interfering with each other. When we first discussed this, my initial thought was that we could perhaps hook the retrieval of orca task info into the existing resource monitoring loop because, if an orca task fails, in theory the resource would never resolve to the desired state, in which case we’d eventually give up and generate a failure event — at which point we could try to obtain more info about the related tasks, as opposed to just blindly polling orca all the time. The disadvantage is that it might take much longer to detect the need to ask orca for details. 

At first it seemed cleaner to have a separate loop/schedule, but I’m afraid it might end up adding a lot more complexity.

Thoughts?

Reply:
Thanks for your comments @luispollo ! It makes complete sense to add here some mechanisms for distributed systems. I will refactor the repository to use PeriodicallyCheckedRepository.

Reply:
@asher would love to get your feedback on the following:
For other classes that implement `PeriodicallyCheckedRepository`, there are 2 tables - the actual data table and the XXX_last_checked table. Do you think it makes sense to add another table to hold the task_last_checked here? Or maybe just use a different mechanism to find the correct tasks?

Reply:
I think a different mechanism would be preferable since tasks should be short lived. I think it would make sense to hold a lock on running `OrcaTaskMonitorAgent` so that only one keel instance is checking tasks at a time, but not at the individual task level.

I wrote a simple sql based agent lock implementation for this purpose in orca: https://github.com/spinnaker/orca/blob/master/orca-sql/src/main/kotlin/com/netflix/spinnaker/orca/notifications/SqlNotificationClusterLock.kt

The same pattern would work well here. That class could optionally be moved to `kork` with names made more generic to avoid reimplementing.